add FileSource activation/deactivation to MapSnapshotter

tobrun opened the pull request:
This PR fixes the bug of the MapSnapshot occasionally not showing up.

ivovandongen commented on platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/storage/FileSource.java:
```diff
@@ -135,6 +134,11 @@ public void activate() {
   }
 
   public void deactivate() {
+    if (activeCounter == 0) {
```
It seems that we should deal with these multiple deactivations at the source (MapSnapshotter). Otherwise, if `activate` / `deactivate` is not called in pairs, we have no solid state tracking for the activation state and it could be deactivated before all dependents are actually ready for deactivation.

tobrun commented:
@ivovandongen you are correct, was able to find a couple of use-cases were the initial approach would go horribly wrong. Been revisiting activation/deactivation though not very happy with the required changes. The underlying issue is that MapSnapshotter is reusable + it doesn't have a lifecycle (if this was the case we could just hook into the C++ destructor and we would be done). 

To handle all use-cases "gracefully" with out current exposed API, I moved ref counting code to the C++ side and added ref counting itself to mapsnapshotter (to handle reuse/multiple requests). 

The edge case I have been focusing on is loading multiple maps with the MapSnapshotter and killing the activity before MapSnapshotter is done:
 - If the user uses cancel on MapSnapshotter in onStop, the filesource is immediately paused.
 - If the users isn't using cancel on MapSnapshotter, the filesource will be paused when the snapshotter has finished generating snapshots in the background (even if the callback has been destroyed). 

Do you see a better solution for this issue (with taking semver in account)?

ivovandongen commented on platform/android/src/snapshotter/map_snapshotter.cpp:
```diff
@@ -79,16 +83,24 @@ void MapSnapshotter::start(JNIEnv&) {
     snapshotCallback.reset();
-}
 
+    if (activationCounter > 0) {
```
Can we simplify this to a boolean in the MapSnapshotter? Don't think we need an actual counter in this case right?

tobrun replied:
This is needed because the MapSnapshotter can be reused for generating mulitple maps. When start is called we increase by one, when we receive a callback we decrease by one [here](https://github.com/mapbox/mapbox-gl-native/pull/10556/files#diff-71ad6c747ca48b85bf4f30842aa0a2a7R86). If the end developer calls into cancel (eg. as part of destroying the activity) we are resetting the counter. A boolean doesn't give us an indication if we have ongoing requests, eg. generating multiple maps in one activity would stop working after generating one bitmap.

ivovandongen commented on platform/android/src/file_source.cpp:
```diff
@@ -63,11 +63,18 @@ void FileSource::setResourceTransform(jni::JNIEnv& env, jni::Object<FileSource::
 void FileSource::resume(jni::JNIEnv&) {
-    fileSource->resume();
+    counter++;
+    if (counter==1 && wasPaused) {
```
Do we need to check both `counter` and `wasPaused`? The latter seems redundant.

ivovandongen replied:
Also, spaces (`==`)

tobrun replied:
it seems that it's only valid to call resume if the filesource was paused first (else the assertion will fail in core). I reworked the code to use remove wasPaused and wrapped the counter in an optional instead.

ivovandongen commented on platform/android/src/file_source.hpp:
```diff
@@ -54,6 +54,8 @@ class FileSource {
     static void registerNative(jni::JNIEnv&);
 
 private:
+    int counter = 0;
```
Use a more descriptive name, like `activationCounter` you used elsewhere?

ivovandongen commented on platform/android/src/snapshotter/map_snapshotter.cpp:
```diff
@@ -79,16 +83,24 @@ void MapSnapshotter::start(JNIEnv&) {
             javaPeer->Call(*_env, onSnapshotReady, mapSnapshot);
         }
+        activationCounter--;
+        if(activationCounter==0){
```
Spaces

ivovandongen commented on platform/android/src/file_source.cpp:
```diff
@@ -63,11 +63,18 @@ void FileSource::setResourceTransform(jni::JNIEnv& env, jni::Object<FileSource::
 void FileSource::pause(jni::JNIEnv&) {
-    fileSource->pause();
+    counter--;
+    if (counter==0) {
```
spaces

tobrun commented:
@ivovandongen you were right on the Snapshotter, the counter has been replaced with a bool. 
This PR is ready for re-review.

tobrun commented:
The unrelated formatting changes have been fixed in 00795d7. Going to merge once CI approves.
